File: app/src/main/java/de/markusfisch/android/shadereditor/fragment/CropImageFragment.java
package de.markusfisch.android.shadereditor.fragment;

import android.app.Activity;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.view.LayoutInflater;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.core.view.MenuProvider;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Lifecycle;

import java.util.concurrent.Executors;

import de.markusfisch.android.shadereditor.R;
import de.markusfisch.android.shadereditor.activity.AbstractSubsequentActivity;
import de.markusfisch.android.shadereditor.graphics.BitmapEditor;
import de.markusfisch.android.shadereditor.widget.CropImageView;

public class CropImageFragment extends Fragment {
	public interface CropImageViewProvider {
		CropImageView getCropImageView();
	}

	public static final String IMAGE_URI = "image_uri";
	private static boolean inProgress = false;
	private CropImageView cropImageView;
	private View progressView;
	private Uri imageUri;
	private Bitmap bitmap;

	@Override
	public void onCreate(Bundle state) {
		super.onCreate(state);
	}

	@Override
	public View onCreateView(
			@NonNull LayoutInflater inflater,
			ViewGroup container,
			Bundle state) {
		Activity activity = getActivity();
		if (activity == null) {
			return null;
		}
		activity.setTitle(R.string.crop_image);

		try {
			cropImageView = ((CropImageViewProvider) activity)
					.getCropImageView();
		} catch (ClassCastException e) {
			throw new ClassCastException(activity.toString() +
					" must implement " +
					"CropImageFragment.CropImageViewProvider");
		}

		Bundle args = getArguments();
		if (args == null ||
				(imageUri = args.getParcelable(IMAGE_URI)) == null) {
			abort(activity);
			return null;
		}

		View view = inflater.inflate(
				R.layout.fragment_crop_image,
				container,
				false);
		progressView = view.findViewById(R.id.progress_view);

		// Make cropImageView in activity visible (again).
		cropImageView.setVisibility(View.VISIBLE);

		view.findViewById(R.id.crop).setOnClickListener(v -> cropImage());

		requireActivity().addMenuProvider(new MenuProvider() {
			@Override
			public void onCreateMenu(@NonNull android.view.Menu menu,
					@NonNull MenuInflater menuInflater) {
				menuInflater.inflate(R.menu.fragment_crop_image, menu);
			}

			@Override
			public boolean onMenuItemSelected(@NonNull MenuItem menuItem) {
				if (menuItem.getItemId() == R.id.rotate_clockwise) {
					rotateClockwise();
					return true;
				}
				return false;
			}
		}, getViewLifecycleOwner(), Lifecycle.State.RESUMED);

		return view;
	}

	@Override
	public void onResume() {
		super.onResume();
		loadBitmapAsync();
	}

	private void loadBitmapAsync() {
		final Activity activity = getActivity();
		if (activity == null || inProgress) {
			return;
		}

		inProgress = true;
		progressView.setVisibility(View.VISIBLE);
		Handler handler = new Handler(Looper.getMainLooper());
		Executors.newSingleThreadExecutor().execute(() -> {
			Bitmap b = BitmapEditor.getBitmapFromUri(
					activity,
					imageUri,
					1024);
			handler.post(() -> {
				inProgress = false;
				progressView.setVisibility(View.GONE);

				if (b == null) {
					abort(activity);
					return;
				}

				if (isAdded()) {
					if (bitmap != null) {
						bitmap.recycle();
					}
					// `b` is non-premultiplied. We need a premultiplied one for display.
					if (b.hasAlpha() && !b.isPremultiplied()) {
						// Create a premultiplied copy for display.
						Bitmap premultipliedBitmap = b.copy(b.getConfig(), true);
						premultipliedBitmap.setPremultiplied(true);
						b.recycle();
						bitmap = premultipliedBitmap;
					} else {
						bitmap = b;
					}
					cropImageView.setImageBitmap(bitmap);
				}
			});
		});
	}

	private void abort(Activity activity) {
		Toast.makeText(
				activity,
				R.string.cannot_pick_image,
				Toast.LENGTH_SHORT).show();
		activity.finish();
	}

	private void cropImage() {
		AbstractSubsequentActivity.addFragment(
				getParentFragmentManager(),
				Sampler2dPropertiesFragment.newInstance(
						imageUri,
						cropImageView.getNormalizedRectInBounds(),
						cropImageView.getImageRotation()));

		bitmap.recycle();
		bitmap = null;

		cropImageView.setImageBitmap(null);
		cropImageView.setVisibility(View.GONE);
	}

	private void rotateClockwise() {
		cropImageView.setImageRotation(
				(cropImageView.getImageRotation() + 90) % 360);
	}
}

File: app/src/main/java/de/markusfisch/android/shadereditor/fragment/CropImageFragment.kt
package de.markusfisch.android.shadereditor.fragment

import android.app.Activity
import android.graphics.Bitmap
import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.core.os.BundleCompat
import androidx.core.view.MenuProvider
import androidx.fragment.app.Fragment
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import de.markusfisch.android.shadereditor.R
import de.markusfisch.android.shadereditor.activity.AbstractSubsequentActivity
import de.markusfisch.android.shadereditor.graphics.BitmapEditor
import de.markusfisch.android.shadereditor.widget.CropImageView
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

class CropImageFragment : Fragment() {

    fun interface CropImageViewProvider {
        fun getCropImageView(): CropImageView
    }

    companion object {
        const val IMAGE_URI = "image_uri"
        private var inProgress = false
    }

    private lateinit var cropImageView: CropImageView
    private lateinit var progressView: View
    private lateinit var imageUri: Uri
    private var bitmap: Bitmap? = null

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?
    ): View? {
        val activity = activity ?: return null
        activity.title = getString(R.string.crop_image)

        cropImageView = (activity as? CropImageViewProvider)?.getCropImageView()
            ?: throw ClassCastException("$activity must implement CropImageViewProvider")

        imageUri =
            BundleCompat.getParcelable(requireArguments(), IMAGE_URI, Uri::class.java) ?: run {
                abort(activity)
                return null
            }

        val view = inflater.inflate(R.layout.fragment_crop_image, container, false)
        progressView = view.findViewById(R.id.progress_view)

        // Make cropImageView in activity visible again
        cropImageView.visibility = View.VISIBLE

        view.findViewById<View>(R.id.crop).setOnClickListener { cropImage() }

        activity.addMenuProvider(object : MenuProvider {
            override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
                menuInflater.inflate(R.menu.fragment_crop_image, menu)
            }

            override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
                if (menuItem.itemId == R.id.rotate_clockwise) {
                    rotateClockwise()
                    return true
                }
                return false
            }
        }, viewLifecycleOwner, Lifecycle.State.RESUMED)

        return view
    }

    override fun onResume() {
        super.onResume()
        loadBitmapAsync()
    }

    private fun loadBitmapAsync() {
        val activity = activity ?: return
        if (inProgress) return

        inProgress = true
        progressView.visibility = View.VISIBLE

        // Launch background task to load the bitmap
        lifecycleScope.launch {
            val b = withContext(Dispatchers.IO) {
                BitmapEditor.getBitmapFromUri(activity, imageUri, 1024)
            }

            inProgress = false
            progressView.visibility = View.GONE

            if (b == null) {
                abort(activity)
            } else if (isAdded) {
                bitmap = b
                cropImageView.setImageBitmap(b)
            }
        }
    }

    private fun abort(activity: Activity) {
        Toast.makeText(activity, R.string.cannot_pick_image, Toast.LENGTH_SHORT).show()
        activity.finish()
    }

    private fun cropImage() {
        val uri = imageUri
        AbstractSubsequentActivity.addFragment(
            parentFragmentManager, Sampler2dPropertiesFragment.newInstance(
                uri, cropImageView.normalizedRectInBounds, cropImageView.imageRotation
            )
        )
        bitmap?.recycle()
        bitmap = null

        cropImageView.setImageBitmap(null)
        cropImageView.visibility = View.GONE
    }

    private fun rotateClockwise() {
        cropImageView.imageRotation = (cropImageView.imageRotation + 90) % 360
    }
}